kpi-tracker/src/main.rs

```diff
@@ -79,117 +83,400 @@ struct BlockDetails {
+        let row = db.query_opt(&self.get_latest_height, &[]).await?;
+        if let Some(row) = row {
+            let raw = row.try_get::<_, i64>(0)?;
+            Ok(Some((raw as u64).into()))
+        } else {
+            Ok(None)
+        }
+    }
 }
 
-/// Represents a relation between a contract and a transaction
-#[derive(Debug, Hash, PartialEq, PartialOrd, Ord, Eq)]
-struct TransactionContractRelation {
-    contract:    ContractAddress,
-    transaction: TransactionHash,
+/// Holds [`tokio_postgres::Client`] to query the database and
+/// [`PreparedStatements`] which can be executed with the client.
+struct DBConn {
+    client:   tokio_postgres::Client,
+    prepared: PreparedStatements,
 }
 
-type BlocksTable = HashMap<BlockHash, BlockDetails>;
-type AccountsTable = HashMap<CanonicalAccountAddress, AccountDetails>;
-type AccountTransactionsTable = HashMap<TransactionHash, TransactionDetails>;
-type ContractModulesTable = HashMap<ModuleRef, ContractModuleDetails>;
-type ContractInstancesTable = HashMap<ContractAddress, ContractInstanceDetails>;
-type TransactionsAccountsTable = HashSet<TransactionAccountRelation>;
-type TransactionsContractsTable = HashSet<TransactionContractRelation>;
-
-/// This is intended as a in-memory DB, which follows the same schema as the
-/// final DB will follow.
-struct DB {
-    /// Table containing all blocks queried from node.
-    blocks: BlocksTable,
-    /// Table containing all accounts created on chain, along with accounts
-    /// present in genesis block
-    accounts: AccountsTable,
-    /// Table containing all account transactions finalized on chain.
-    account_transactions: AccountTransactionsTable,
-    /// Table containing all smart contract modules deployed on chain.
-    contract_modules: ContractModulesTable,
-    /// Table containing all smart contract instances created on chain.
-    contract_instances: ContractInstancesTable,
-    /// Table containing relations between accounts and transactions.
-    transaction_account_relations: TransactionsAccountsTable,
-    /// Table containing relations between contract instances and transactions.
-    transaction_contract_relations: TransactionsContractsTable,
+impl DBConn {
+    /// Create new `DBConn` from `tokio_postgres::config::Config`. If
+    /// `try_create_tables` is true, database tables are created using
+    /// `/resources/schema.sql`.
+    async fn create(
+        conn_string: tokio_postgres::config::Config,
+        try_create_tables: bool,
+    ) -> anyhow::Result<Self> {
+        let (client, connection) = conn_string
+            .connect(NoTls)
+            .await
+            .context("Could not create database connection")?;
+
+        tokio::spawn(async move {
+            if let Err(e) = connection.await {
+                log::error!("Connection error: {}", e);
+            }
+        });
+
+        if try_create_tables {
+            let create_statements = include_str!("../resources/schema.sql");
```

Comment:
i'd prefer if this was in a constant and not inline here. It makes it more visible.

Reply:
Can't seem to do that. It's probably the same reason as what's given to the issue posted here:
https://github.com/rust-lang/rust/issues/69133